### PR TITLE
feat: pass job envs to /etc/environment for next reload

### PR DESCRIPTION
**Pull request by PrashantRaj18198**
## Desc
Job specific ENVs are not restored on snapshot based VM but this should be the default behaviour. This PR handles this.

## Changes
- Saves environment variable of current job to `/etc/environment`. We trim out the GitHub and WarpBuild specific and some ID based ENVs.
- Logs the environment variables that are written to `/etc/environment` for more clarity.
- Removes `PATH` from .env file in `/runner` since this overrides the PATH coming from `/etc/environment`.

### CODE REVIEW

File: src/snapshotter/snapshotter.ts [hunk 1]
```diff
@@ -62,68 +62,94 @@ export class Snapshotter {
...
+  if grep -q "^\${key}=" /etc/environment; then
+    sudo sed -i "s|^\${key}=.*|\${key}=\${value}|" /etc/environment
+  else
+    printf '%s=%q\n' "$key" "$value" | sudo tee -a /etc/environment > /dev/null
+  fi
+done
+
+echo "Logging environment variables in /etc/environment"
+echo "--------------------------------"
+cat /etc/environment
```
Comment (prajjwaldimri) [inline src/snapshotter/snapshotter.ts:88]: Wouldn't this be logged in CI logs? Env vars will have sensitive info.

File: src/snapshotter/snapshotter.ts [hunk 2]
```diff
@@ -62,68 +62,94 @@ export class Snapshotter {
...
+echo "--------------------------------"
+
+# Remove PATH from /runner/.env
+sed -i '/^PATH=/d' /runner/.env
```
Comment (prajjwaldimri) [inline src/snapshotter/snapshotter.ts:92]: IIRC, this PATH is used by GitHub runner when spinning up the job. Will this be correctly populated next time the new VM spins up?

File: src/snapshotter/snapshotter.ts [hunk 3]
```diff
@@ -62,68 +62,94 @@ export class Snapshotter {
...
+
+echo "Logging github runner .env file"
+echo "--------------------------------"
+cat /runner/.env
```
Comment (prajjwaldimri) [inline src/snapshotter/snapshotter.ts:96]: Does this file have sensitive logs?

File: src/snapshotter/snapshotter.ts [hunk 4]
```diff
@@ -62,68 +62,94 @@ export class Snapshotter {
...
 
-    const cleanupScript = `
+    const presaveScript = `
 #!/bin/bash
 
 set -e
 
+echo "Saving environment variables to /etc/environment"
+
+printenv | grep -v '^GITHUB_' | grep -v '^WARP_' | grep -v '^ACTIONS_' | grep -v '^INPUT_' | grep -v '^WARPBUILD_' | grep -v '^RUNNER_' | grep -v '^_' | grep -v '^USER$' | grep -v '^INVOCATION_ID$' | grep -v '^SYSTEMD_EXEC_PID$' | grep -v '^SUDO_USER$' | grep -v '^SUDO_GID$' | grep -v '^SUDO_UID$' | grep -v '^SHELL$' | grep -v '^SHLVL$' | grep -v '^XDG_' | grep -v '^JOURNAL_STREAM$' | grep -v '^HOME$' | sort | while IFS= read -r line; do
```
Comment (prajjwaldimri) [inline src/snapshotter/snapshotter.ts:76]: Is this list documented? We should write the reason behind removing these vars.

File: src/snapshotter/snapshotter.ts (hunk 2 above)
Comment (PrashantRaj18198) [reply src/snapshotter/snapshotter.ts:92]: Yes. PATH is being set properly

File: src/snapshotter/snapshotter.ts (hunk 3 above)
Comment (PrashantRaj18198) [reply src/snapshotter/snapshotter.ts:96]: Nope. It just have the PATH env.

File: src/snapshotter/snapshotter.ts (hunk 1 above)
Comment (PrashantRaj18198) [reply src/snapshotter/snapshotter.ts:88]: It shouldn't matter that much, GitHub hides the sensitive info from logs.

**Review by prajjwaldimri — APPROVED** [review]
LGTM 🏰

**Pull request merged**